modules/core/src/lib/layer-state.ts

```diff
@@ -0,0 +1,47 @@
+import ComponentState from '../lifecycle/component-state';
+
+import type Layer from './layer';
+import type AttributeManager from './attribute/attribute-manager';
+import {LayerProps} from './layer-props';
+
+export default class LayerState<T extends LayerProps> extends ComponentState<T> {
+  attributeManager: AttributeManager;
+  needsRedraw: boolean;
+  subLayers: any;
+  usesPickingColorCache: boolean;
+
+  constructor({attributeManager, layer}: {attributeManager: AttributeManager; layer: Layer}) {
+    super(layer);
+    this.attributeManager = attributeManager;
+    this.needsRedraw = true;
+    this.subLayers = null; // reference to sublayers rendered in a previous cycle
+    this.usesPickingColorCache = false;
+  }
+
+  get layer(): Layer {
+    return this.component as Layer;
+  }
+
+  set layer(layer: Layer) {
+    this.component = layer;
+  }
+
+  protected _fetch(propName, url: string) {
```

Comment:
Nit: These are new methods right. Generally best to avoid refactoring code while adding types, given that changes drown in the noise? Also I would generally recommend avoiding adding methods on a state object and keep it as "pure" as possible.

Reply:
These methods access props (`fetch`, `onDataLoad` etc) that are only defined for Layer, not the base Component class.

> I would generally recommend avoiding adding methods on a state object

Not sure what you are suggesting here. This class has always been gigantic and have very complex internal and external APIs since it's conception.